android/app/src/main/java/com/breez/client/plugins/breez/breezlib/GoogleAuthenticator.java

```diff
@@ -152,6 +157,105 @@ public boolean validateAccessTokenAllowingPrompt() {
         }
     }
 
+    private boolean verifyCredentialHasWriteAccess(String token, boolean allowRecover) {
+        Log.d(TAG, "verifyCredentialHasWriteAccess token = " + token);
+        if (token == null || token.isEmpty()) {
+            return false;
+        }
+
+        try {
+            Drive driveService = new Drive.Builder(
+                    new NetHttpTransport.Builder().build(),
+                    new GsonFactory(),
+                    request -> {
+                    })
+                    .setHttpRequestInitializer(req -> req.getHeaders().setAuthorization("Bearer " + token))
+                    .setApplicationName("Breez").build();
+            FileList result = driveService.files().list().setSpaces("appDataFolder").execute();
+            Log.d(TAG, "verifyCredentialHasWriteAccess result = " + result);
+        } catch (Exception e) {
+            if (e instanceof GoogleJsonResponseException && ((GoogleJsonResponseException) e).getStatusCode() == 401) {
+                if (allowRecover) {
+                    Log.d(TAG, "verifyCredentialHasWriteAccess failed with 401, trying to sign in to recovery write permission");
+                    return recoverUnauthorizedAccess();
+                } else {
+                    Log.w(TAG, "verifyCredentialHasWriteAccess failed with 401, but not allowed to recover");
+                    return false;
+                }
+            } else {
+                Log.w(TAG, "verifyCredentialHasWriteAccess failed", e);
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean recoverUnauthorizedAccess() {
```

Comment:
The logic seems to work but the code seems a bit complicated, probably because of this duplication.
Let's think how to make it more compact and better reused.

Reply:
I did compact the tries/catches and other minor improvements, let me know if you see more things that we can improve to make it more readable.